module/zfs/zfs_log.c

```diff
@@ -558,68 +602,97 @@ zfs_log_write(zilog_t *zilog, dmu_tx_t *tx, int txtype,
 	else
 		write_state = WR_NEED_COPY;
 
+	/*
+	 * A WR_COPIED record must fit entirely in one log block.
+	 * Large writes can use WR_NEED_COPY, which the ZIL will
+	 * split into multiple records across several log blocks
+	 * if necessary.
+	 */
+	uint64_t max_wr_copied_lr_length = zil_max_copied_data(zilog);
+	if (write_state == WR_COPIED && resid > max_wr_copied_lr_length)
+		write_state = WR_NEED_COPY;
+
 	if ((fsync_cnt = (uintptr_t)tsd_get(zfs_fsyncer_key)) != 0) {
 		(void) tsd_set(zfs_fsyncer_key, (void *)(fsync_cnt - 1));
 	}
 
-	while (resid) {
-		itx_t *itx;
-		lr_write_t *lr;
-		itx_wr_state_t wr_state = write_state;
-		ssize_t len = resid;
+	boolean_t sync = B_TRUE;
+	if (!(ioflag & (O_SYNC | O_DSYNC)) && (zp->z_sync_cnt == 0) &&
+	    (fsync_cnt == 0))
+		sync = B_FALSE;
 
-		/*
-		 * A WR_COPIED record must fit entirely in one log block.
-		 * Large writes can use WR_NEED_COPY, which the ZIL will
-		 * split into multiple records across several log blocks
-		 * if necessary.
-		 */
-		if (wr_state == WR_COPIED &&
-		    resid > zil_max_copied_data(zilog))
-			wr_state = WR_NEED_COPY;
-		else if (wr_state == WR_INDIRECT)
-			len = MIN(blocksize - P2PHASE(off, blocksize), resid);
-
-		itx = zil_itx_create(txtype, sizeof (*lr) +
-		    (wr_state == WR_COPIED ? len : 0));
-		lr = (lr_write_t *)&itx->itx_lr;
-
-		/*
-		 * For WR_COPIED records, copy the data into the lr_write_t.
-		 */
-		if (wr_state == WR_COPIED) {
-			int err;
-			DB_DNODE_ENTER(db);
-			err = dmu_read_by_dnode(DB_DNODE(db), off, len, lr + 1,
-			    DMU_READ_NO_PREFETCH);
-			if (err != 0) {
-				zil_itx_destroy(itx);
-				itx = zil_itx_create(txtype, sizeof (*lr));
-				lr = (lr_write_t *)&itx->itx_lr;
-				wr_state = WR_NEED_COPY;
-			}
-			DB_DNODE_EXIT(db);
+	itx_t *itx;
+	if (write_state == WR_NEED_COPY) {
+		itx = zfs_log_write_itx_create(
+		    0,
+		    max_wr_copied_lr_length,
+		    write_state,
+		    zp,
+		    off,
+		    resid,
+		    sync,
+		    callback,
+		    callback_data,
+		    NULL);
```

Comment:
Is there any reason for this formatting to be that long?  Imagine you have 80x25 screen.

Reply:
Haha my personal taste is that it's better to have each argument on a separate line for such long parameter lists because it makes (future) diffs where the function signature is changed easier to review. Not a hill I'll die on though.